task/mgmt-gateway/src/mgs_handler.rs

```diff
@@ -206,19 +204,69 @@ impl SpHandler for MgsHandler {
             return Err(ResponseError::RequestUnsupportedForComponent);
         }
 
-        // TODO serial console access should require auth; for now, receiving
-        // serial console data implicitly attaches us
+        if self.attached_serial_console_mgs.is_some() {
+            return Err(ResponseError::SerialConsoleAlreadyAttached);
+        }
+
+        // TODO: Add some kind of auth check before allowing a serial console
+        // attach. https://github.com/oxidecomputer/hubris/issues/723
         self.attached_serial_console_mgs = Some((sender, port));
+        self.serial_console_write_offset = 0;
+        self.usart.from_rx_offset = 0;
+        Ok(())
+    }
 
-        // TODO-cleanup This rejects the entire message if it doesn't all fit in
-        // our buffer; we could accept part of the data if some of it fits.
-        // Needs cooperation from MGS.
-        if self.usart.tx_buffer_remaining_capacity() >= data.len() {
-            self.usart.tx_buffer_append(data);
-            Ok(())
-        } else {
-            Err(ResponseError::Busy)
+    fn serial_console_write(
+        &mut self,
+        sender: SocketAddrV6,
+        port: SpPort,
+        mut offset: u64,
+        mut data: &[u8],
+    ) -> Result<u64, ResponseError> {
+        ringbuf_entry!(Log::MgsMessage(MgsMessage::SerialConsoleWrite {
+            offset,
+            length: data.len() as u16
+        }));
+
+        // TODO: Add some kind of auth check before allowing a serial console
+        // attach. https://github.com/oxidecomputer/hubris/issues/723
+        //
+        // As a temporary measure, we can at least ensure that we only allow
+        // writes from the attached console.
+        if Some((sender, port)) != self.attached_serial_console_mgs {
+            return Err(ResponseError::SerialConsoleNotAttached);
         }
+
+        // Have we already received some or all of this data? (MGS may resend
+        // packets that for which it hasn't received our ACK.)
+        if self.serial_console_write_offset > offset {
+            let skip = self.serial_console_write_offset - offset;
+            // Have we already seen _all_ of this data? If so, just reply that
+            // we're ready for the data that comes after it.
+            if skip >= data.len() as u64 {
+                return Ok(offset + data.len() as u64);
```

Comment:
Should we return `self.serial_console_write_offset` here instead?

e.g. if 
```
self.serial_console_write_offset = 128
offset = 64
data.len() = 32
```
we want the host to start sending us data at offset 128, not 96.

(or will the host get confused that we've managed to write more data than it thinks it's sent us?)

Reply:
> (or will the host get confused that we've managed to write more data than it thinks it's sent us?)

Exactly this, at least as currently implemented. In practice we should never see `skip > data.len()` here, unless something has gone horribly wrong, because the host should never go backwards in time, and only moves forward when we ack its messages. We may see `skip == data.len()` if our ack gets lost and the host resends the same packet it most recently sent, though.